File: c++/nitf/tests/test_writer_3.cpp
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 *
 * (C) Copyright 2004 - 2009, General Dynamics - Advanced Information Systems
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program; if not, If not,
 * see <http://www.gnu.org/licenses/>.
 *
 */

#include <import/nitf.hpp>
#include <iostream>
#include <string>

/* *********************************************************************
 ** This test tests the round-trip process of taking an input NITF
 ** file and writing it to a new file. This includes writing the image
 ** segments (headers, extensions, and image data). This is an example
 ** of how users can write the image data to their NITF file
 ** ********************************************************************/

class JunkWriteHandler : public nitf::WriteHandler
{
public:
    JunkWriteHandler(){}
    ~JunkWriteHandler(){}

    void write(nitf::IOHandle handle) throw (nitf::NITFException)
    {
        //yup, does nothing...

        handle.write("01234567890", 11);
    }

};


nitf::Record doRead(const std::string& inFile);

std::string makeBandName(const std::string& rootFile, int imageNum, int bandNum)
{
    int pos = rootFile.find_last_of("/\\");
    std::ostringstream os;
    os << rootFile.substr(pos + 1) << "__" << imageNum << "_band_" << bandNum;
    std::string newFile = os.str();

    while ((pos = newFile.find(".")) != std::string::npos)
        newFile.replace(pos, 1, "_");
    newFile += ".man";
    return newFile;
}

nitf::ImageSource* setupBands(int nbands, int imageNum, const std::string& inRootFile)
{
    nitf::ImageSource* iSource = new nitf::ImageSource();
    for (int i = 0; i < nbands; i++)
    {
        std::string inFile = makeBandName(inRootFile, imageNum, i);
        nitf::IOHandle sourceHandle(inFile);
        iSource->addBand(new nitf::FileSource(sourceHandle, 0, 1, 0));
    }
    return iSource;
}

void doWrite(nitf::Record record, const std::string& inRootFile, const std::string& outFile)
{
    nitf::Writer writer;
    nitf::IOHandle output_io(outFile, NITF_ACCESS_WRITEONLY, NITF_CREATE);
    writer.prepare(output_io, record);

    int numImages = record.getHeader().getNumImages();
    nitf::ListIterator end = record.getImages().end();
    nitf::ListIterator iter = record.getImages().begin();
    nitf_Error error;

    for (int i = 0; i < numImages && iter != end; ++i, ++iter)
    {
        nitf::ImageSegment imseg;
        imseg = *iter;
        int nbands = imseg.getSubheader().getNumImageBands();
        nitf::ImageWriter* iWriter = writer.newImageWriter(i);
        //nitf::WriteHandler *iWriter = new JunkWriteHandler();
        //writer.setImageWriteHandler(i, iWriter);

        nitf::ImageSource* iSource = setupBands(nbands, i, inRootFile);
        iWriter->attachSource(iSource, true);
    }
    writer.write();
    output_io.close();
}

int main(int argc, char **argv)
{
    try
    {
        /*  Check argv and make sure we are happy  */
        if (argc != 3)
        {
            std::cout << "Usage: %s <input-file> <output-file> \n" << argv[0] << std::endl;
            exit(EXIT_FAILURE);
        }
        nitf::Record record = doRead(argv[1]);
        doWrite(record, argv[1], argv[2]);
    }
    catch (except::Throwable & t)
    {
        std::cout << t.getMessage() << std::endl;
    }
}


void manuallyWriteImageBands(nitf::ImageSegment & segment,
                             const std::string& imageName,
                             nitf::ImageReader& deserializer,
                             int imageNumber)
{
    int padded;

    nitf::Uint32 nBits = segment.getSubheader().getNumBitsPerPixel();
    nitf::Uint32 nBands = segment.getSubheader().getNumImageBands();
    nitf::Uint32 xBands = segment.getSubheader().getNumMultispectralImageBands();
    nBands += xBands;

    nitf::Uint32 nRows = segment.getSubheader().getNumRows();
    nitf::Uint32 nColumns = segment.getSubheader().getNumCols();
    //nitf::Uint32 nBlocksPerCol = segment.getSubheader().getNumBlocksPerCol();
    //nitf::Uint32 nPixPerVertBlock = segment.getSubheader().getNumPixelsPerVertBlock();


    //one row at a time
    size_t subWindowSize = nColumns * NITF_NBPP_TO_BYTES(nBits);

    std::cout << "NBANDS -> " << nBands << std::endl
        << "XBANDS -> " << xBands << std::endl
        << "NROWS -> " << nRows << std::endl
        << "NCOLS -> " << nColumns << std::endl
        << "PVTYPE -> " << segment.getSubheader().getPixelValueType().toString() << std::endl
        << "NBPP -> " << segment.getSubheader().getNumBitsPerPixel().toString() << std::endl
        << "ABPP -> " << segment.getSubheader().getActualBitsPerPixel().toString() << std::endl
        << "PJUST -> " << segment.getSubheader().getPixelJustification().toString() << std::endl
        << "IMODE -> " << segment.getSubheader().getImageMode().toString() << std::endl
        << "NBPR -> " << segment.getSubheader().getNumBlocksPerRow().toString() << std::endl
        << "NBPC -> " << segment.getSubheader().getNumBlocksPerCol().toString() << std::endl
        << "NPPBH -> " << (int)segment.getSubheader().getNumPixelsPerHorizBlock() << std::endl
        << "NPPBV -> " << (int)segment.getSubheader().getNumPixelsPerVertBlock() << std::endl
        << "IC -> " << segment.getSubheader().getImageCompression().toString() << std::endl
        << "COMRAT -> " << segment.getSubheader().getCompressionRate().toString() << std::endl;

    nitf::Uint8** buffer = new nitf::Uint8*[nBands];
    nitf::Uint32* bandList = new nitf::Uint32[nBands];
    for (nitf::Uint32 band = 0; band < nBands; band++)
        bandList[band] = band;

    nitf::SubWindow subWindow;
    subWindow.setStartCol(0);
    subWindow.setNumRows(1);
    subWindow.setNumCols(nColumns);

    // necessary ?
    nitf::DownSampler* pixelSkip = new nitf::PixelSkip(1, 1);
    subWindow.setDownSampler(pixelSkip);
    subWindow.setBandList(bandList);
    subWindow.setNumBands(nBands);

    assert(buffer);
    for (int i = 0; i < nBands; i++)
    {
        buffer[i] = new nitf::Uint8[subWindowSize];
        assert(buffer[i]);
    }

    std::vector<nitf::IOHandle> handles;
    //make the files
    for (int i = 0; i < nBands; i++)
    {
        std::string name = makeBandName(imageName, imageNumber, i);
        nitf::IOHandle toFile(name, NITF_ACCESS_WRITEONLY, NITF_CREATE);
        handles.push_back(toFile);
    }

    //nitf::Uint32 lastBlockRows = nPixPerVertBlock - (nBlocksPerCol * nPixPerVertBlock - nRows);
    //size_t lastBlockSize = nColumns * NITF_NBPP_TO_BYTES(nBits) * lastBlockRows;

    //read all row blocks and write to disk
    for (int i = 0; i < nRows; ++i)
    {
        subWindow.setStartRow(i);

        //if (i == nBlocksPerCol - 1)
        //    subWindow.setNumRows(lastBlockRows);

        deserializer.read(subWindow, buffer, &padded);
        for (int j = 0; j < nBands; j++)
        {
            //if (i == nBlocksPerCol - 1)
            //    handles[j].write((const char*)buffer[j], lastBlockSize);
            //else
            handles[j].write((const char*)buffer[j], subWindowSize);
        }
    }

    //close output handles
    for (int i = 0; i < nBands; i++)
        handles[i].close();

    /* free buffers */
    for (int i = 0; i < nBands; i++)
        delete [] buffer[i];
    delete [] buffer;
    delete [] bandList;
    delete pixelSkip;
}


nitf::Record doRead(const std::string& inFile)
{
    nitf::Reader reader;
    nitf::IOHandle io(inFile);
    nitf::Record record = reader.read(io);

    /*  Set this to the end, so we'll know when we're done!  */
    nitf::ListIterator end = record.getImages().end();
    nitf::ListIterator iter = record.getImages().begin();
    for (int count = 0, numImages = record.getHeader().getNumImages();
            count < numImages && iter != end; ++count, ++iter)
    {
        nitf::ImageSegment imageSegment = *iter;
        nitf::ImageReader deserializer = reader.newImageReader(count);
        std::cout << "Writing image " << count << "..." << std::endl;

        /*  Write the thing out  */
        manuallyWriteImageBands(imageSegment, inFile, deserializer, count);
        std::cout << "done.\n" << std::endl;
    }

    return record;
}

File: externals/nitro_2.10.2/modules/java/nitf/src/samples/TestWriter3.java
/* =========================================================================
 * This file is part of NITRO
 * =========================================================================
 * 
 * (C) Copyright 2004 - 2010, MDA Information Systems LLC
 *
 * NITRO is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public 
 * License along with this program; if not, If not, 
 * see <http://www.gnu.org/licenses/>.
 *
 */

import java.io.File;
import java.io.IOException;

import nitf.FileSource;
import nitf.IOHandle;
import nitf.ImageReader;
import nitf.ImageSegment;
import nitf.ImageSource;
import nitf.ImageWriter;
import nitf.NITFException;
import nitf.Reader;
import nitf.Record;
import nitf.SubWindow;
import nitf.Writer;

/**
 * This test is a Java version of the C test_writer_3 test case. <p/> This test
 * reads in the given NITF file, then loops over the image segments, and writes
 * out each band to a separate file. <p/> Finally, it creates a BandSource for
 * each of the band files, and writes out a final NITF. None of the fields are
 * updated in the output NITF file, so if all works well, a diff of the input
 * NITF file and output NITF file should produce no results. <p/> *NOTE -
 * Currently, only writing headers and image segments works properly. If you try
 * testing with an image that has other segments, including TREs, you will
 * probably produce a file that diffs with the original. Future releases will
 * have a more complete writing capability.
 */
public class TestWriter3
{
    public static void main(String[] args) throws NITFException, IOException
    {

        if (args.length != 2)
        {
            System.out.println("Usage: java " + TestWriter3.class.getName()
                    + " <input-file> <output-file>");
            System.exit(1);
        }

        Record record = doRead(args[0]);
        record.getHeader().print(System.out);
        doWrite(record, args[0], args[1]);
    }

    /**
     * This constructs a name for the raw image data for each band
     * 
     * @param rootFile
     * @param imageNum
     * @param bandNum
     * @return
     */
    private static String makeBandName(String rootFile, int imageNum,
            int bandNum)
    {
        File file = new File(rootFile);
        String name = file.getName();
        String fileName = name + "__" + imageNum + "_band_" + bandNum;
        fileName = fileName.replaceAll("[.]", "_");
        fileName += ".man";
        return new File(file.getParentFile(), fileName).getPath();
    }

    /**
     * Reads the NITF file, and calls manuallyWriteImageBands for each image
     * 
     * @param inFile
     * @return
     * @throws nitf.NITFException
     */
    private static Record doRead(String inFile) throws nitf.NITFException
    {
        int numImages;

        Reader reader = new Reader();

        IOHandle io = new IOHandle(inFile, IOHandle.NITF_ACCESS_READONLY,
                IOHandle.NITF_OPEN_EXISTING);

        Record record = reader.read(io);
        numImages = record.getHeader().getNumImages().getIntData();

        System.out.println("record.getHeader() = " + record.getHeader());

        ImageSegment[] images = record.getImages();
        for (int i = 0; i < images.length; i++)
        {
            ImageSegment imageSegment = images[i];
            ImageReader deserializer = reader.getNewImageReader(i);
            System.out.println("Writing image " + i);
            /* Write the thing out */
            manuallyWriteImageBands(imageSegment, inFile, deserializer, i);
            System.out.println("Done");
        }

        return record;
    }

    /**
     * This function writes out each band to its own file
     * 
     * @param segment
     * @param imageName
     * @param deserializer
     * @param imageNumber
     * @throws NITFException
     */
    private static void manuallyWriteImageBands(ImageSegment segment,
            String imageName, ImageReader deserializer, int imageNumber)
            throws NITFException
    {
        int nBits = segment.getSubheader().getNumBitsPerPixel().getIntData();
        int nBands = segment.getSubheader().getNumImageBands().getIntData();
        int xBands = segment.getSubheader().getNumMultispectralImageBands()
                .getIntData();
        nBands += xBands;

        int nRows = segment.getSubheader().getNumRows().getIntData();
        int nColumns = segment.getSubheader().getNumCols().getIntData();
        int subimageSize = nRows * nColumns * ((nBits - 1) / 8 + 1);
        System.out.println("Window: " + subimageSize);

        System.out.println("NBANDS -> " + nBands);
        System.out.println("NROWS -> " + nRows);
        System.out.println("NCOLS -> " + nColumns);
        System.out.println("PVTYPE -> "
                + segment.getSubheader().getPixelValueType().getStringData());

        byte[][] buffer = new byte[nBands][subimageSize];
        int band = 0;

        SubWindow subimage = new SubWindow();
        subimage.setStartCol(0);
        subimage.setStartRow(0);
        subimage.setNumCols(nColumns);
        subimage.setNumRows(nRows);
        int[] bandList = new int[nBands];
        for (int i = 0; i < bandList.length; i++)
        {
            bandList[i] = i;
        }
        subimage.setBandList(bandList);
        subimage.setNumBands(nBands);

        System.out.println("About to read");

        deserializer.read(subimage, buffer);
        System.out.println("just read");
        for (int i = 0; i < nBands; i++)
        {
            String fileName = makeBandName(imageName, imageNumber, i);
            IOHandle toFile = new IOHandle(fileName,
                    IOHandle.NITF_ACCESS_WRITEONLY, IOHandle.NITF_CREATE);

            toFile.write(buffer[i], subimageSize);
            toFile.close();
        }
        System.out.println("done");
    }

    /**
     * This writes the final NITF file
     * 
     * @param record
     * @param inRootFile
     * @param outFile
     * @throws NITFException
     */
    private static void doWrite(Record record, String inRootFile, String outFile)
            throws NITFException
    {
        IOHandle output_io = new IOHandle(outFile,
                IOHandle.NITF_ACCESS_WRITEONLY, IOHandle.NITF_CREATE);

        Writer writer = new Writer();
        boolean b = writer.prepare(record, output_io);
        System.out.println("b = " + b);

        ImageSegment[] images = record.getImages();
        for (int i = 0; i < images.length; i++)
        {
            ImageSegment imageSegment = images[i];

            // just for giggles, lets add a LUT for this
            /*
             * BandInfo info = imageSegment.getSubheader().getBandInfo()[i]; if
             * (info.getLookupTable() == null) { byte[] data = new byte[256];
             * for (int j = 0; j < 256; ++j) data[i] = (byte)j; LookupTable lut =
             * new LookupTable(1, 256, data); info.setLookupTable(lut); }
             */

            int numImageBands = imageSegment.getSubheader().getNumImageBands()
                    .getIntData();

            ImageWriter iWriter = writer.getNewImageWriter(i);

            ImageSource iSource = setupBands(numImageBands, i, inRootFile);

            iWriter.attachSource(iSource);
        }

        boolean wroteOK = writer.write();
        System.out.println("wroteOK = " + wroteOK);

        output_io.close();
    }

    /**
     * Creates an imagesource from each of the bands
     * 
     * @param nbands
     * @param imageNum
     * @param inRootFile
     * @return
     * @throws NITFException
     */
    private static ImageSource setupBands(int nbands, int imageNum,
            String inRootFile) throws NITFException
    {
        ImageSource iSource = new ImageSource();
        for (int i = 0; i < nbands; i++)
        {
            String inFile = makeBandName(inRootFile, imageNum, i);
            IOHandle sourceHandle = new IOHandle(inFile,
                    IOHandle.NITF_ACCESS_READONLY, IOHandle.NITF_OPEN_EXISTING);

            FileSource bandSource = new FileSource(sourceHandle, 0, 1, 0);
            iSource.addBand(bandSource);
        }
        return iSource;
    }
}
